Added Blog Post (#6436)
ChilliCream/graphql-platform: website/src/blog/2023-08-15-fusion/2023-08-15-fusion.md

## Expectations

Our belief is that distributed GraphQL services — or composite GraphQL services — should not only be straightforward to set up but also seamlessly integrate with the diverse range of CI/CD tools, schema registries, composition utilities, and gateways that enterprises might prefer. The current landscape should not dictate the choice of tools but provide flexibility.

Up to this point, the GraphQL landscape has lacked an open specification tailored for distributed setups – a framework designed from the ground up for extensibility and integration with diverse toolchains. We envisioned a platform where tools from various vendors could effortlessly work in tandem, ensuring that developers and enterprises never feel constrained by their technical choices.

## Lets share and compete

> Late last year ChilliCream and The Guild met in Paris and discussed their approaches twoards distributed GraphQL. It became very clear that both companies were solving similar problems and we decided to join forces on this project. ChilliCream would provide the initial work on the Fusion spec and implementation while The Guild would start specifying their work on Mash and Open API and help shape the initial Fusion spec.

Reviewer: ```suggestion
Late last year ChilliCream and [The Guild](https://the-guild.dev/) met in Paris and discussed their approaches twoards distributed GraphQL. It became very clear that both companies were solving similar problems and we decided to join forces on this project. ChilliCream would provide the initial work on the Fusion spec and implementation while The Guild would start specifying their work on the [GraphQL Mesh Gateway](https://the-guild.dev/graphql/mesh) with [OpenAPI support](https://the-guild.dev/graphql/mesh/docs/handlers/openapi) and [Hive](https://the-guild.dev/graphql/hive) as an open schema registry to help shape the initial Fusion spec.
```

> Late last year ChilliCream and The Guild met in Paris and discussed their approaches twoards distributed GraphQL. It became very clear that both companies were solving similar problems and we decided to join forces on this project. ChilliCream would provide the initial work on the Fusion spec and implementation while The Guild would start specifying their work on Mash and Open API and help shape the initial Fusion spec.

Reviewer 2: ```suggestion
Late last year ChilliCream and The Guild met in Paris and discussed their approaches towards distributed GraphQL. It became very clear that both companies were solving similar problems and we decided to join forces on this project. ChilliCream would provide the initial work on the Fusion spec and implementation while The Guild would start specifying their work on Mash and Open API and help shape the initial Fusion spec.
```

> Today, we are thrilled to unveil GraphQL-Fusion an open specification under the MIT license. This initiative empowers everyone to craft tools and solutions centered around distributed GraphQL services. Complementing this announcement, we're also introducing Hot Chocolate Fusion, an early implementation of the GraphQL-Fusion spec draft.

Reviewer: ```suggestion
Today, we are thrilled to unveil GraphQL-Fusion an open specification under the **MIT license**. This initiative empowers everyone to craft tools and solutions centered around distributed GraphQL services. Complementing this announcement, we're also introducing Hot Chocolate Fusion, an early implementation of the GraphQL-Fusion spec draft.
```

Reviewer: Should you also mention here the working group?

Reviewer: and are you ok with adding something like "And The Guild commitment to integrate Fusion into their existing gateway solutions" or something like that?

Alternatively, if your subgraphs implement the Relay conventions, such as the global object identity convention, the schema composition will detect this and incorporate it into the Fusion graph document. This can optimize your query planning, for example, by utilizing the node field.

```graphql
extend type User
  @resolver(
    subgraph: "User",
    select: "{ node(id: $User_id) { ... on User { ... User } } }",
    arguments: [ { name: "User_id", type: "ID!" } ])
  @resolver(
    subgraph: "User",
    select: "{ nodes(ids: $User_id) { ... on User { ... User } } }",
    arguments: [ { name: "User_id", type: "[ID!]!" } ],
    kind: "BATCH_BY_KEY") {
}
```

While using the `node` field to fetch entity data is straightforward for exposing the `node` fields to the Gateway, we found it necessary to equip the Fusion gateway with data sharding capabilities. This is essentially the ability to dispatch a query at runtime to a subgraph based on user-provided data. This can be applied to simple tasks like the node field, but can also be harnessed to isolate data partitions by region or any other discriminants you desire.

## Going Further

> While the subgraph inference of the schema composition is quite powerfull there are a lot of cases where we can further by declaring semntics of a schema. We do not need to integrate such annotations into our subgraph schema directly but can pass an additional GraphQL documents into the schema composition that hold type extensions with additional directive annotations.

Author: ```suggestion
While the subgraph inference of the schema composition is quite powerful there are a lot of cases where we can go further by declaring semantics of a schema. We do not need to integrate such annotations into our subgraph schema directly but can pass an additional GraphQL documents into the schema composition that hold type extensions with additional directive annotations.
```

```graphql
type Product {
  sku: String!
  name: String!
  dimension: ProductDimension
  deliveryEstimate(zip: String!): Int!
}
```

We can express this by using the `@require` directive and refer to the required information relatively to the `Product` type.

```graphql
type Product {
  deliveryEstimate(
    zip: String!,
    width: Float! @require(field: "dimension { width }"),
    height: Float! @require(field: "dimension { height }")) : Int!
}
```

We also could design that slightly differently and introduce a input to our subgraph that represents the required data we need.

```graphql
input ProductDimensionInput {
  width: Float!,
  height: Float!
}

type Product {
  deliveryEstimate(
    zip: String!,
    dimension: ProductDimensionInput! @require(field: "dimension")): Int!
}
```

The outcome will stay the same and we get this nice API for our user and the query planner will resolve the requiered data under the hood.

![Banana Cake Pop - Query Plan Viewer](bcp-2.png)

Again, this brings actual clarity to your subgraph as the field is very clear what it needs and also becomes easily testable.

## Reshaping things

> When we rethink a bit our shipping subgraph we actually should realize that the `deliveryEstimate` does not reall need to be on a product type as the argument has clear requirements which are expressed by it's field arguments. Instead of having the field `deliveryEstimate` on the `Product` type itself it could very well be exposed through the `Query` type. At least in the context of our subgraph.

Reviewer 3: ```suggestion
When we rethink a bit our shipping subgraph we actually should realize that the `deliveryEstimate` does not really need to be on a product type as the argument has clear requirements which are expressed by its field arguments. Instead of having the field `deliveryEstimate` on the `Product` type itself, it could be exposed through the `Query` type. At least in the context of our subgraph.
```

```graphql
type Query {
  estimateShipping(zip: String!, width: Float!, height: Float!): Int!
}
```

> In the case our subgraph is isolated and does not fully integrate with our though for model that we want to expose on the gateway we can also reshape it.

Reviewer 3: ```suggestion
In the case, our subgraph is isolated and does not fully integrate with our though for the model that we want to expose on the gateway we can also reshape it.
```

> All the annotations can be put into separate graphql document providing the extending metadata. First lets make the whole query type private, we actually do not want to include anything by default form this subgraph.

Reviewer 3: ```suggestion
All the annotations can be put into separate graphql documents providing the extending metadata. First lets make the whole query type private, we actually do not want to include anything by default form this subgraph.
```

The Open Packaging Convention is an open standard provided by Microsoft, used for everything from Word Documents to VSCode extension packages. Simply put, think of it as a ZIP container with metadata and relations between its artifacts. The GraphQL-Fusion convention contains the mandatory Fusion Graph document. This document is all that's needed to run and configure a Gateway implementing the core specification.

Additionally, it contains all subgraph schema documents and the publicly exposed Gateway schema. Having all these artifacts in one place gives us a single artifact that we can pass from the schema composition in a CI/CD pipeline to the schema registry, and from there to the gateway. We have customers already using this with their custom solutions for distributing the configuration from their deployment pipeline to their Gateway or by using our Cloud Services. This ensures that neither your consumers nor your gateway experiences any interruptions.

Besides these standard components, it also allows Gateway implementers to store custom configurations to specify GraphQL WAF settings and more.

[SCHEMA of the Fusion config]

## Conclusion

> We are still working on GraphQL-Fusion, but you can try it today with the Hot Chocolate Fusion Gateway. We plan to release the GraphQL-Fusion spec as it matures later this year under the MIT license. While the current iterations are primarily concerned with GraphQL, The Guild has begun specifying the Open API to GraphQL transformation they've developed in Mesh, and we'll integrate this as it becomes available. Additionally, there's much more from the query plan engine and composition we're eager to showcase, such as field requirements, transforming subgraphs to better integrate into your overall graph, even manual annotations instead of inference, and the fact that you can take any Apollo Subgraph and seamlessly incorporate it into the schema composition. This allows it to work in orchestration with other non-Apollo Federation subgraphs. Currently, we're all hands on deck working on a step-by-step tutorial, which will guide you through every detail, from a simple setup to a full-fledged enterprise solution. This includes features like schema change tracking and build-time errors that inform you of potential issues even before your first subgraph commit.

Reviewer: I would give the part where you can take any Apollo Subgraph and seamlessly migrate it to Fusion its own headline and paragraph
Something around seamless migration

Author: I will ... also want to add an example snippet of that.

Today, we are thrilled to unveil GraphQL-Fusion, an open specification under the **MIT license**. This initiative empowers everyone to craft tools and solutions centered around distributed GraphQL services. Complementing this announcement, we're also introducing Hot Chocolate Fusion, an early implementation of the GraphQL-Fusion spec draft.

## A new way to distribute GraphQL schema components

GraphQL Fusion presents a fresh approach to streamlining the complexities of assembling distributed schemas.

At its heart, GraphQL Fusion pivots around two foundational principles: schema composition and query planning.

> But before delving into these concepts, it's essential to retrace our steps. Let's revisit the challenges surfaced when people first tried GraphQL as a Gateway for constructing their GraphQL servers over REST APIs.

Reviewer 3: ```suggestion
But before delving into these concepts, it's essential to retrace our steps. Let's revisit the challenges that surfaced when people first tried GraphQL as a Gateway for constructing their GraphQL servers over REST APIs.
```

Author: I think the that is not needed in this sentence.

Additionally, it contains all subgraph schema documents, the publicly exposed Gateway schema, and composition settings the user has opted into. Having all these artifacts in one place gives us a single artifact that we can pass on from the schema composition in a CI/CD pipeline to the schema registry and from there to the actual gateway. We have customers already using this with their custom solutions for distributing the configuration from their deployment pipeline to their gateway or by using our Cloud Services ([Banana Cake Pop](https://eat.bananacakepop.com)). Besides these standard artifacts included in the package, it also allows Gateway implementers to store custom configurations to specify GraphQL WAF settings and more.

![Simple Deployment Pipeline](pipeline-1.png)

## Apollo Federation

> We recognize that some of you may have opted into Apollo Federation as a solution, and that is why we designed the Fusion schema composition so that it can compose any Apollo Federation subgraph into a Fusion subgraph. This allows you to seamlessly migration from an Apollo Federation setup to a GraphQL-Fusion setup without the need to rewrite everything.

Reviewer 3: ```suggestion
We recognize that some of you may have opted into Apollo Federation as a solution, and that is why we designed the Fusion schema composition so that it can compose any Apollo Federation subgraph into a Fusion subgraph. This allows you to seamlessly migrate from an Apollo Federation setup to a GraphQL-Fusion setup without the need to rewrite everything.
```
